# Importing libraries

In [2]:
import numpy as np # for performing mathematical calculations behind ML algorithms
import matplotlib.pyplot as plt # for visualization
import pandas as pd # for handling and cleaning the dataset
import seaborn as sns # for visualization

# Importing the dataset

In [3]:
dataset = pd.read_csv('CarPrice_Assignment.csv')

# Analyzing the data

In [4]:
dataset.head()

car_ID  symboling                   CarName fueltype aspiration doornumber   
0       1          3        alfa-romero giulia      gas        std        two  \
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize   
0  convertible        rwd          front       88.6  ...         130  \
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg   
0        mpfi       3.47    2.68              9.0        111     5000      21  \
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

In [5]:
dataset.tail()

car_ID  symboling          CarName fueltype aspiration doornumber   
200     201         -1  volvo 145e (sw)      gas        std       four  \
201     202         -1      volvo 144ea      gas      turbo       four   
202     203         -1      volvo 244dl      gas        std       four   
203     204         -1        volvo 246   diesel      turbo       four   
204     205         -1      volvo 264gl      gas      turbo       four   

    carbody drivewheel enginelocation  wheelbase  ...  enginesize  fuelsystem   
200   sedan        rwd          front      109.1  ...         141        mpfi  \
201   sedan        rwd          front      109.1  ...         141        mpfi   
202   sedan        rwd          front      109.1  ...         173        mpfi   
203   sedan        rwd          front      109.1  ...         145         idi   
204   sedan        rwd          front      109.1  ...         141        mpfi   

     boreratio  stroke compressionratio horsepower  peakrpm citympg   
200       3.78    3.15              9.5        114     5400      23  \
201       3.78    3.15              8.7        160     5300      19   
202       3.58    2.87              8.8        134     5500      18   
203       3.01    3.40             23.0        106     4800      26   
204       3.78    3.15              9.5        114     5400      19   

     highwaympg    price  
200          28  16845.0  
201          25  19045.0  
202          23  21485.0  
203          27  22470.0  
204          25  22625.0  

[5 rows x 26 columns]

In [6]:
# Numerical/Statistical analysis of the dataset
dataset.describe()

car_ID   symboling   wheelbase   carlength    carwidth   carheight   
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000  \
mean   103.000000    0.834146   98.756585  174.049268   65.907805   53.724878   
std     59.322565    1.245307    6.021776   12.337289    2.145204    2.443522   
min      1.000000   -2.000000   86.600000  141.100000   60.300000   47.800000   
25%     52.000000    0.000000   94.500000  166.300000   64.100000   52.000000   
50%    103.000000    1.000000   97.000000  173.200000   65.500000   54.100000   
75%    154.000000    2.000000  102.400000  183.100000   66.900000   55.500000   
max    205.000000    3.000000  120.900000  208.100000   72.300000   59.800000   

        curbweight  enginesize   boreratio      stroke  compressionratio   
count   205.000000  205.000000  205.000000  205.000000        205.000000  \
mean   2555.565854  126.907317    3.329756    3.255415         10.142537   
std     520.680204   41.642693    0.270844    0.313597          3.972040   
min    1488.000000   61.000000    2.540000    2.070000          7.000000   
25%    2145.000000   97.000000    3.150000    3.110000          8.600000   
50%    2414.000000  120.000000    3.310000    3.290000          9.000000   
75%    2935.000000  141.000000    3.580000    3.410000          9.400000   
max    4066.000000  326.000000    3.940000    4.170000         23.000000   

       horsepower      peakrpm     citympg  highwaympg         price  
count  205.000000   205.000000  205.000000  205.000000    205.000000  
mean   104.117073  5125.121951   25.219512   30.751220  13276.710571  
std     39.544167   476.985643    6.542142    6.886443   7988.852332  
min     48.000000  4150.000000   13.000000   16.000000   5118.000000  
25%     70.000000  4800.000000   19.000000   25.000000   7788.000000  
50%     95.000000  5200.000000   24.000000   30.000000  10295.000000  
75%    116.000000  5500.000000   30.000000   34.000000  16503.000000  
max    288.000000  6600.000000   49.000000   54.000000  45400.000000

In [7]:
# Dimensions of dataset
print('There are ',dataset.shape[0],'rows and ',dataset.shape[1],'columns in the dataset.') 

There are  205 rows and  26 columns in the dataset.


In [8]:
# Here we are trying to check that whether this dataset has any duplicate value or not
print('There are',dataset.duplicated().sum(),'duplicate values in the dateset.') #using duplicated() pre-defined function

There are 0 duplicate values in the dateset.


In [9]:
# Check for NULL values
dataset.isnull().sum()

car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

In [10]:
# Schema of dataset
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

# Getting counts of each categories of columns with Datatype=object

In [11]:
print(dataset.CarName.value_counts())
print(dataset.fueltype.value_counts())
print(dataset.aspiration.value_counts())
print(dataset.doornumber.value_counts())
print(dataset.carbody.value_counts())
print(dataset.drivewheel.value_counts())
print(dataset.enginelocation.value_counts())
print(dataset.enginetype.value_counts())
print(dataset.cylindernumber.value_counts())
print(dataset.fuelsystem.value_counts())

CarName
toyota corona           6
toyota corolla          6
peugeot 504             6
subaru dl               4
mitsubishi mirage g4    3
                       ..
mazda glc 4             1
mazda rx2 coupe         1
maxda glc deluxe        1
maxda rx3               1
volvo 246               1
Name: count, Length: 147, dtype: int64
fueltype
gas       185
diesel     20
Name: count, dtype: int64
aspiration
std      168
turbo     37
Name: count, dtype: int64
doornumber
four    115
two      90
Name: count, dtype: int64
carbody
sedan          96
hatchback      70
wagon          25
hardtop         8
convertible     6
Name: count, dtype: int64
drivewheel
fwd    120
rwd     76
4wd      9
Name: count, dtype: int64
enginelocation
front    202
rear       3
Name: count, dtype: int64
enginetype
ohc      148
ohcf      15
ohcv      13
dohc      12
l         12
rotor      4
dohcv      1
Name: count, dtype: int64
cylindernumber
four      159
six        24
five       11
eight       5
two         4
three 

# 1) Converting String Values to Int using Label Encoder(For Carnames as there are many types of car)

# 2) Converting the other object types to int with replace 

In [17]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
dataset['CarName'] = le.fit_transform(dataset['CarName'])
dataset.replace({'fueltype':{'gas':0,'diesel':1}},inplace=True)
dataset.replace({'aspiration':{'std':0,'turbo':1}},inplace=True)
dataset.replace({'doornumber':{'four':0,'two':1}},inplace=True)
dataset.replace({'carbody':{'sedan':0,'hatchback':1,'wagon':2,'hardtop':3,'convertible':4}},inplace=True)
dataset.replace({'drivewheel':{'fwd':0,'rwd':1,'4wd':2}},inplace=True)
dataset.replace({'enginelocation':{'front':0,'rear':1}},inplace=True)
dataset.replace({'enginetype':{'ohc':0,'ohcf':1,'ohcv':2,'dohc':3,'l':4,'rotor':5,'dohcv':6}},inplace=True)
dataset.replace({'cylindernumber':{'four':0,'six':1,'five':2,'eight':3,'two':4,'three':5,'twelve':6}},inplace=True)
dataset.replace({'fuelsystem':{'mpfi':0,'2bbl':1,'idi':2,'1bbl':3,'spdi':4,'4bbl':5,'mfi':6,'spfi':7}},inplace=True)
dataset.head()


car_ID  symboling  CarName  fueltype  aspiration  doornumber  carbody   
0       1          3        2         0           0           1        4  \
1       2          3        3         0           0           1        4   
2       3          1        1         0           0           1        1   
3       4          2        4         0           0           0        0   
4       5          2        5         0           0           0        0   

   drivewheel  enginelocation  wheelbase  ...  enginesize  fuelsystem   
0           1               0       88.6  ...         130           0  \
1           1               0       88.6  ...         130           0   
2           1               0       94.5  ...         152           0   
3           0               0       99.8  ...         109           0   
4           2               0       99.4  ...         136           0   

   boreratio  stroke  compressionratio  horsepower  peakrpm  citympg   
0       3.47    2.68               9.0         111     5000       21  \
1       3.47    2.68               9.0         111     5000       21   
2       2.68    3.47               9.0         154     5000       19   
3       3.19    3.40              10.0         102     5500       24   
4       3.19    3.40               8.0         115     5500       18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

# Model Development

In [19]:
print(x)

     car_ID  symboling  CarName  fueltype  aspiration  doornumber  carbody   
0         1          3        2         0           0           1        4  \
1         2          3        3         0           0           1        4   
2         3          1        1         0           0           1        1   
3         4          2        4         0           0           0        0   
4         5          2        5         0           0           0        0   
..      ...        ...      ...       ...         ...         ...      ...   
200     201         -1      139         0           0           0        0   
201     202         -1      138         0           1           0        0   
202     203         -1      140         0           0           0        0   
203     204         -1      142         1           1           0        0   
204     205         -1      143         0           1           0        0   

     drivewheel  enginelocation  wheelbase  ...  cylindernumber

In [20]:
print(y)

0      13495.0
1      16500.0
2      16500.0
3      13950.0
4      17450.0
        ...   
200    16845.0
201    19045.0
202    21485.0
203    22470.0
204    22625.0
Name: price, Length: 205, dtype: float64


# Preparing the Train and Test Data

In [21]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2,random_state=0)
print(x.shape,x_train.shape,x_test.shape)

(205, 25) (164, 25) (41, 25)


# Training through different Models

# 1) Linear Regression

In [22]:
from sklearn.linear_model import LinearRegression
model1=LinearRegression()
model1.fit(x_train,y_train)

LinearRegression()

In [23]:
test_data_prediction1=model1.predict(x_test)

In [24]:
print(test_data_prediction1)

[ 4697.51933023 17162.03167431 17319.83517961  6141.3059653
 10293.53524806 12892.6125781   7667.91481306  6628.49434384
 18656.5530908   6842.50824631 17607.50657628 26145.68791624
 10631.62633617 14280.33074465  7761.68120978 12558.59577176
  9113.3947309  21776.62515956  8708.44827702  5944.75670982
 11071.31427217 16678.72835074  9933.27876987 14412.88796844
 19738.48672023  7051.80513473  5705.41751383 19111.56935124
  6499.61252419 11271.90878836  8443.88387801 10733.12013728
 16916.87986259  9574.94665805  6373.0618868  25730.66570449
  6930.44809111 13793.35816856  7125.24687609 36012.52425764
  7571.82324762]


In [25]:
testing_data_model_score = model1.score(x_test, y_test)
print("Model Score/Performance on Testing data",testing_data_model_score)

training_data_model_score = model1.score(x_train, y_train)
print("Model Score/Performance on Training data",training_data_model_score)

Model Score/Performance on Testing data 0.8197374446024239
Model Score/Performance on Training data 0.9253844853561265


In [26]:
df = pd.DataFrame(data={'Predicted value':test_data_prediction1,'Actual Value':y_test})
df

Predicted value  Actual Value
52       4697.519330        6795.0
181     17162.031674       15750.0
5       17319.835180       15250.0
18       6141.305965        5151.0
188     10293.535248        9995.0
170     12892.612578       11199.0
76       7667.914813        5389.0
154      6628.494344        7898.0
104     18656.553091       17199.0
33       6842.508246        6529.0
12      17607.506576       20970.0
129     26145.687916       31400.5
55      10631.626336       10945.0
66      14280.330745       18344.0
45       7761.681210        8916.5
169     12558.595772        9989.0
130      9113.394731        9295.0
7       21776.625160       18920.0
37       8708.448277        7895.0
152      5944.756710        6488.0
80      11071.314272        9959.0
111     16678.728351       15580.0
131      9933.278770        9895.0
171     14412.887968       11549.0
179     19738.486720       15998.0
138      7051.805135        5118.0
156      5705.417514        6938.0
113     19111.569351       16695.0
161      6499.612524        8358.0
89      11271.908788        5499.0
183      8443.883878        7975.0
193     10733.120137       12290.0
125     16916.879863       22018.0
173      9574.946658        8948.0
92       6373.061887        6849.0
16      25730.665704       41315.0
189      6930.448091       11595.0
136     13793.358169       18150.0
22       7125.246876        6377.0
74      36012.524258       45400.0
44       7571.823248        8916.5

# Model Evaluation(Linear Regression):

In [27]:
from sklearn.metrics import r2_score
r2Score = r2_score(test_data_prediction1, y_test)
print("R2 score of model is :" ,r2Score*100)

R2 score of model is : 68.44863463333095


In [28]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(test_data_prediction1,y_test)
print("Mean Absolute Error is :" ,mae)

Mean Absolute Error is : 2472.6491092636625


Conclusion

So, the mean absolute error is 2472.6491092636625. Therefore our predicted value can be 2472.6491092636625 units greater than or less than the actual value.

# 2) RFC (Random Forest Classifier)

In [29]:
from sklearn.ensemble import RandomForestRegressor
model2= RandomForestRegressor(n_estimators=100, random_state=0)
model2.fit(x_train, y_train)

RandomForestRegressor(random_state=0)

In [30]:
test_data_prediction2=model2.predict(x_test)

In [31]:
print(test_data_prediction2)

[ 5949.26    16348.26167 14447.3      6031.       9645.03    14529.51
  5960.81     7599.47    18111.27501  6713.2     19363.57    35765.48
 12419.68    14392.47     6357.65    13317.46    12210.14    18405.93167
  8576.26     6565.21     9395.88    13744.64    10141.14    14913.13
 17241.71334  7281.35     7686.84    14879.11     7849.53     7715.57
  8412.88    12957.66    15934.15334 10126.87     7117.38    32761.36
  9470.04    16971.3      6073.85    37381.8      6117.26   ]


In [32]:
testing_data_model_score = model2.score(x_test, y_test)
print("Model Score/Performance on Testing data",testing_data_model_score)

training_data_model_score = model2.score(x_train, y_train)
print("Model Score/Performance on Training data",training_data_model_score)

Model Score/Performance on Testing data 0.9048564739162297
Model Score/Performance on Training data 0.9907108903232087


In [33]:
df = pd.DataFrame(data={'Predicted value':test_data_prediction2,'Actual Value':y_test})
df

Predicted value  Actual Value
52        5949.26000        6795.0
181      16348.26167       15750.0
5        14447.30000       15250.0
18        6031.00000        5151.0
188       9645.03000        9995.0
170      14529.51000       11199.0
76        5960.81000        5389.0
154       7599.47000        7898.0
104      18111.27501       17199.0
33        6713.20000        6529.0
12       19363.57000       20970.0
129      35765.48000       31400.5
55       12419.68000       10945.0
66       14392.47000       18344.0
45        6357.65000        8916.5
169      13317.46000        9989.0
130      12210.14000        9295.0
7        18405.93167       18920.0
37        8576.26000        7895.0
152       6565.21000        6488.0
80        9395.88000        9959.0
111      13744.64000       15580.0
131      10141.14000        9895.0
171      14913.13000       11549.0
179      17241.71334       15998.0
138       7281.35000        5118.0
156       7686.84000        6938.0
113      14879.11000       16695.0
161       7849.53000        8358.0
89        7715.57000        5499.0
183       8412.88000        7975.0
193      12957.66000       12290.0
125      15934.15334       22018.0
173      10126.87000        8948.0
92        7117.38000        6849.0
16       32761.36000       41315.0
189       9470.04000       11595.0
136      16971.30000       18150.0
22        6073.85000        6377.0
74       37381.80000       45400.0
44        6117.26000        8916.5

# Model Evaluation(RFC):

In [34]:
from sklearn.metrics import r2_score
r2Score = r2_score(test_data_prediction2, y_test)
print("R2 score of model is :" ,r2Score*100)

R2 score of model is : 86.91583033259086


In [35]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(test_data_prediction2,y_test)
print("Mean Absolute Error is :" ,mae)

Mean Absolute Error is : 1867.480853902439


Conclusion

So, the mean absolute error is 1867.480853902439. Therefore our predicted value can be 1867.480853902439
 units greater than or less than the actual value.

# 3) XGBRegressor(Extreme Gradient Boosting)

In [36]:
from xgboost import XGBRegressor
model = XGBRegressor()
model.fit(x_train,y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [37]:
test_data_prediction=model.predict(x_test)

In [38]:
print(test_data_prediction)

[ 6062.315  16436.965  12483.024   8607.949  10038.932  13368.531
  6023.8916  6885.746  18634.178   6272.4756 21076.326  37178.215
 11846.98   14568.255   6397.986  12205.556  12342.02   18807.652
  8994.833   6147.6235  9143.627  16102.957   9818.865  13829.381
 17682.25    7088.1714  8510.371  14525.625   7641.9727  7493.895
  8311.096  12140.209  15899.997  10271.542   6862.458  32657.924
  9992.617  18561.186   6130.741  38056.652   6271.61  ]


In [39]:
testing_data_model_score = model.score(x_test, y_test)
print("Model Score/Performance on Testing data",testing_data_model_score)

training_data_model_score = model.score(x_train, y_train)
print("Model Score/Performance on Training data",training_data_model_score)

Model Score/Performance on Testing data 0.90493839732248
Model Score/Performance on Training data 0.9999999439625338


In [40]:
df = pd.DataFrame(data={'Predicted value':test_data_prediction,'Actual Value':y_test})
df

Predicted value  Actual Value
52       6062.314941        6795.0
181     16436.964844       15750.0
5       12483.024414       15250.0
18       8607.949219        5151.0
188     10038.931641        9995.0
170     13368.531250       11199.0
76       6023.891602        5389.0
154      6885.746094        7898.0
104     18634.177734       17199.0
33       6272.475586        6529.0
12      21076.326172       20970.0
129     37178.214844       31400.5
55      11846.980469       10945.0
66      14568.254883       18344.0
45       6397.985840        8916.5
169     12205.555664        9989.0
130     12342.019531        9295.0
7       18807.652344       18920.0
37       8994.833008        7895.0
152      6147.623535        6488.0
80       9143.626953        9959.0
111     16102.957031       15580.0
131      9818.865234        9895.0
171     13829.380859       11549.0
179     17682.250000       15998.0
138      7088.171387        5118.0
156      8510.371094        6938.0
113     14525.625000       16695.0
161      7641.972656        8358.0
89       7493.895020        5499.0
183      8311.095703        7975.0
193     12140.208984       12290.0
125     15899.997070       22018.0
173     10271.541992        8948.0
92       6862.458008        6849.0
16      32657.923828       41315.0
189      9992.617188       11595.0
136     18561.185547       18150.0
22       6130.741211        6377.0
74      38056.652344       45400.0
44       6271.609863        8916.5

# Model Evaluation (XGBRegressor)

In [41]:
from sklearn.metrics import r2_score
r2Score = r2_score(test_data_prediction, y_test)
print("R2 score of model is :" ,r2Score*100)

R2 score of model is : 87.64916430918764


In [42]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(test_data_prediction,y_test)
print("Mean Absolute Error is :" ,mae)

Mean Absolute Error is : 1847.3235280106708


Conclusion

So, the mean absolute error is 1847.3235280106708. Therefore our predicted value can be 1847.3235280106708 units greater than or less than the actual value.